ServerLibrary/Envir/Commands/AdminCommandHandler.cs

```diff
@@ -0,0 +1,57 @@
+﻿using Server.Envir.Commands.Admin;
+using Server.Envir.Commands.Command.Admin;
+using Server.Envir.Commands.Handler;
+using Server.Models;
+using System.Collections.Generic;
+
+namespace Server.Envir.Commands
+{
+
+    public class AdminCommandHandler : AbstractCommandHandler
+    {
+
+        public AdminCommandHandler()
+        {
+            InitializeCommands(new List<UserCommand>() {
+                new ChatBan(),
```

Comment:
Can this be initialised generically based on class/interface inheritance. Something like below

public void InitializeCommands<T>() where T : IUserCommand
{
    List<IUserCommand> commands = new List<IUserCommand>();

    foreach (Type type in AppDomain.CurrentDomain.GetAssemblies()
        .SelectMany(s => s.GetTypes())
        .Where(p => typeof(T).IsAssignableFrom(p) && p.IsClass))
    {
        IUserCommand command = (IUserCommand)Activator.CreateInstance(type);
        commands.Add(command);
    }

    // Do something with the list of commands
    // ...
}

InitializeCommands<IUserCommand>();

Reply:
I understand why you want generic solution for initializing, i'm not sure I understand this approach, this `AppDomain.CurrentDomain.GetAssemblies` is not something I am familiar with. How do you determine, UserCommand vs AdminCommand? I guess I would have to add another layer of abstraction over the command to make it deterministic for IAdminCommand or IUserCommand.

I think you should also use something like a .Select() on the IEnumberable instead of using a foreach around it, it is maybe creating a layer of indirection.

It would also still need to inject the command list to the abstract handler where you put the comment, maybe the InitializeCommands(); at the bottom is not necessary or is wrong but it is confusing me.

Reply:
Yes if UserCommand can cover both admin and user, you'll want to create an interface that the commands use - IPlayerCommand and IAdminCommand for example.

Something like this

```
var assembly = Assembly.GetAssembly(typeof(UserCommand)); // use the assembly that contains UserCommand as a starting point
    var commandTypes = assembly.GetTypes().Where(t => t.IsClass && !t.IsAbstract && t.BaseType == typeof(UserCommand) && typeof(IPlayerCommand).IsAssignableFrom(t));
    var commands = new List<UserCommand>();
    
    foreach (var type in commandTypes)
    {
        var command = (IPlayerCommand)Activator.CreateInstance(type);
        commands.Add((UserCommand)command);
    }
```

I wouldn't bother changing the foreach to a linq query - its no quicker and makes it less readable in my opinion.

Reply:
I started working with that notion in mind, i went with the linq query approach though😰 

Just validating it works.

Reply:
I tried to get it working in a generic way, but the result returned empty.

```
    public abstract class AbstractCommandHandler<CommandType> : IValidatingCommandHandler where CommandType : ICommand
    {

        public readonly List<CommandType> Commands = new List<CommandType>();

        public AbstractCommandHandler()
        {
            this.Commands = Assembly.GetAssembly(typeof(ICommand)).GetTypes()
                                .Where(type => type.IsClass)
                                .Where(type => !type.IsAbstract)
                                .Where(type => type.BaseType == typeof(CommandType))
                                .Where(type => typeof(CommandType).IsAssignableFrom(type))
                                .Select(type => (CommandType)Activator.CreateInstance(type))
                                .ToList();
        }
```

Reply:
Ok I've debugged it to work!

```
Assembly.GetAssembly(typeof(CommandType)).GetTypes()
                .Where(type => type.IsSubclassOf(typeof(AbstractCommand<CommandType>))).ToList();
```

Reply:
How are you differentiating between admin and player commands?

Reply:
It's determined by the generic assignment. 
```
class ClearIPBlocks : AbstractCommand<IAdminCommand>
```

```
public class AdminCommandHandler : AbstractCommandHandler<IAdminCommand>
```

which is then used in the assembly code as `CommandType`

```
public abstract class AbstractCommandHandler<CommandType> : IValidatingCommandHandler 
```

Reply:
I'm still trying to work something out, maybe i'm going down a rabbit hole here.

Reply:
@Suprcode the change has been pushed, maybe you want to take a look and see what you think.